### ISSUE TITLE
[EKS] [request]: Provide Ability to Set Hop Limit in NodeClass Resource

### ISSUE THREAD

**Issue by bdumford-OM1** [aws/containers-roadmap#2498]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
I would like the NodeClass resource from api version `eks.amazonaws.com/v1` to be able to set MetaDataOptions, especially the hop limit (`httpPutResponseHopLimit`).

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I am currently blocked from using an application that requires I use pods that need access to the instance metadata.  There is no way to configure this in the NodeClass resource.  I use to be able to do this in Karpenter's [EC2NodeClass](https://karpenter.sh/docs/concepts/nodeclasses/). 

**Are you currently working around this issue?**
I cannot currently work around this issue.

**Additional context**
None

**Attachments**
None

**Comment by lucdew** [comment]
It is probably the reason why we cannot use the aws fluentbit filter and thus augment the logs with AWS metadata.

On Karpenter it works but in AWS EKS auto-mode the retrieval of the token for the metada service v2 fails.

The following request from the fluentbit (2.31) pod times-out:
`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"`

I checked the EC2 instance started by EKS auto-mode and imdsv2 is required.

**Comment by mikestef9** [comment]
This is an intentional decision of Auto Mode, following security best practices. 

https://docs.aws.amazon.com/eks/latest/best-practices/identity-and-access-management.html#_identities_and_credentials_for_eks_pods_recommendations

If you need to have a pod still access IMDS with Auto Mode, you need to run it as hostNetwork. For example if you install the CW container insights addon from EKS addons, you'll notice the fluent-bit and cw-agent daemonsets run as host network.

**Comment by lucdew** [comment]
Ok for fluentbit not being able to configure the httpPutResponseHopLimit is not a problem.

I ran the fluentbit pods in hostNetwork mode set to true and it fixes the metadata retrieval done by the AWS filter plugin.
I should have noticed because the HOST environment variable that was used as a prefix for the cloudwatch log streams was set to the fluentbit pod names and not the EC2 hostname.

So at least for fluentbit use case, being able to set the put response hop limit is not needed. Thanks for giving the reasons why it cannot be changed.

**Comment by lukens** [comment]
@mikestef9 I'm trying to use the Amazon EFS CSI Driver add-on on an EKS Auto Mode cluster, and am getting an error I believe is caused by this:

```
Failed to fetch Access Points or Describe File System: List Access Points failed: operation error EFS: DescribeAccessPoints, get identity: get credentials: failed to refresh cached credentials, no EC2 IMDS role found, operation error ec2imds: GetMetadata, request canceled, context deadline exceeded
```

I tried setting the default hop-limit to 2 as [described in this user guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-IMDS-new-instances.html#:~:text=For%20Metadata%20response,defaults%20to1.), and deleted the node claims so new instances were created, but they still had a hop limit of 1.

I then tried editing the `efs-csi-node` daemonset to add `hostNetwork: true` to the template spec, but I get `Edit cancelled, no changes made.` (I guess the add-on is stopping me doing this).

Is there anything I can do to get this working with auto mode?

Also, I don't _think_ there is anything I the [considerations](https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html#efs-csi-considerations) for using the `efs-csi` driver that suggests it won't work in an Auto Mode cluster.

**Comment by vnandha** [comment]
> This is an intentional decision of Auto Mode, following security best practices.
> 
> https://docs.aws.amazon.com/eks/latest/best-practices/identity-and-access-management.html#_identities_and_credentials_for_eks_pods_recommendations
> 
> If you need to have a pod still access IMDS with Auto Mode, you need to run it as hostNetwork. For example if you install the CW container insights addon from EKS addons, you'll notice the fluent-bit and cw-agent daemonsets run as host network.

Some of the components require access to metadata like Istio proxy so it can use locality based routing. It would be nice to enable these option for those use cases

**Comment by sdarwin** [comment]
Trying to install the AWS Gateway API Controller on EKS Auto Mode.

- The Pod Identities Agent fails due to a port 80 conflict. This is a known issue "EKS] [BAD-DECISION]: EKS Pod Identity agent daemonset mapped to node-port 80"

- Switch to AWS IAM Roles for Service Accounts (IRSA) instead.  This can't reach the IMDS service presumably due to httpPutResponseHopLimit=1.

So the Gateway API Controller can't be installed?

**Comment by mikestef9** [comment]
You don't need to run the Pod Identity agent if you are running Auto Mode, it's already baked into the compute. And Gateway API controller should support Pod Identity so you shouldn't need IRSA at all.

**Comment by sdarwin** [comment]
> it's baked in
> Gateway API controller should support Pod Identity 

Well, I have tried both methods, Pod Identity and IRSA, per the instructions.  

https://www.gateway-api-controller.eks.aws.dev/dev/guides/deploy/

Currently there are Metadata errors

`
{"level":"fatal","ts":"2025-03-25T00:28:14.881Z","logger":"setup","caller":"runtime/proc.go:272","msg":"init config failed: vpcId is not specified: EC2MetadataError: failed to make EC2Metadata request\n\n\tstatus code: 401, request id: "}
`

In those instructions it doesn't distinguish between which steps are newly baked in, and which must still be run.  

I went through all the Pod Identities Agent steps first.    

At that time, the Gateway API Controller pods failed to launch, with those errors.     The Pod Identity agent failed also, but maybe it's because that is already running in Auto Mode.      

In Auto Mode, nothing stops you (either in the console or the command line) from trying to install an add-on which is now also a systemd process.    That ought to be prevented, right?

**Comment by sdarwin** [comment]
> anti affinity

Interesting. The instructions I used (above) listed `--addon-version v1.0.0-eksbuild.1` which did not exclude `auto` nodes.   A new feature in `v1.3.4-eksbuild.1`

> not need to make any IMDS call.

Ok right.  Yes I will do that in the near term.  (it still feels like a work-around since the IMDS is more convenient.)

**Comment by mikestef9** [comment]
Ah yes only newer versions (introduced post auto mode launch) have the anti affinity. And yes, don't disagree it's a bit of a workaround. For EKS addons, we inject automatically variables like region and cluster name. So if the aws gateway api controller was added as an EKS addon, that would be one way to solve this.

**Comment by sdarwin** [comment]
"EKS addon" +1

**Comment by richie-tt** [comment]
> There is anti affinity in addons like Pod Identity to prevent it from running on Auto Mode compute. If you specify things like region and vpcId in the Helm installation values https://github.com/aws/aws-application-networking-k8s/blob/main/helm/values.yaml#L82 - the controller will not need to make any IMDS call.

I disagree with you. From your comment, I understand that it is safer **not to use IMDS** (with which all the necessary values ​​can be read automatically) and copy them manually.

The IMDS is secure if you use `v2`, which requires the token for authentication, but without the possibility to change the hop limit, the endpoint cannot be reached. 

I think this is the monkey job if you always need to copy the `region` and `vpcID` for Helm, when you have the whole automation in place, why do you want to block access to it?

**Comment by eatmyrust** [comment]
What is the point of having IMDS if it is not possible to use it from containerized workloads? There are always security risks to be evaluated when making configuration changes, but that should be left to the user to decide what their risk tolerance is. Surely it's not more secure to run a pod in host network mode? Maintaining a hop limit of 1 by default makes sense to ensure security for users who do not use IMDS, but I don't understand why this is not an option at all for users who want to utilize IMDS.
